app/src/main/java/nodomain/freeyourgadget/gadgetbridge/contentprovider/HRContentProvider.java

```diff
@@ -140,85 +138,130 @@ public Cursor query(@NonNull Uri uri, String[] projection, String selection, Str
-                return mc;
+                LOG.info("REALTIME");
+                return realtime(projection, selectionArgs);
         }
         return null;
     }
 
+    @Nullable
+    private Cursor devices_list() {
+        MatrixCursor mc;
+        DeviceManager deviceManager = ((GBApplication) (this.getContext())).getDeviceManager();
+        List<GBDevice> l = deviceManager.getDevices();
+        if (l == null) {
+            return null;
+        }
+        LOG.info(String.format("listing %d devices", l.size()));
+
+        mc = new MatrixCursor(HRContentProviderContract.deviceColumnNames);
+        for (GBDevice dev : l) {
+            mc.addRow(new Object[]{dev.getName(), dev.getModel(), dev.getAddress()});
+        }
+        return mc;
+    }
+
+    @NonNull
+    private Cursor activity_start(String[] projection, String[] args) {
+        MatrixCursor mc;
+        this.state = provider_state.CONNECTING;
+
+        GBDevice targetDevice = getDevice((args != null) ? args[0] : "");
+        if (targetDevice != null && targetDevice.isConnected()) {
+            enable_realtime();
+            mc = new MatrixCursor(HRContentProviderContract.activityColumnNames);
+            mc.addRow(new String[]{"OK", "Connected"});
+        } else {
+            GBApplication.deviceService().connect(targetDevice);
+            mc = new MatrixCursor(HRContentProviderContract.activityColumnNames);
+            mc.addRow(new String[]{"OK", "Connecting"});
+        }
+
+        return mc;
+    }
+
+    @NonNull
+    private Cursor activity_stop(String[] projection, String[] args) {
+        MatrixCursor mc;
+        this.state = provider_state.INACTIVE;
+
+        GBApplication.deviceService().onEnableRealtimeSteps(false);
+        GBApplication.deviceService().onEnableRealtimeHeartRateMeasurement(false);
+        mc = new MatrixCursor(HRContentProviderContract.activityColumnNames);
+        mc.addRow(new String[]{"OK", "No error"});
+        punchTimer.cancel();
+        punchTimer = new Timer();
+        return mc;
+    }
+
+
+    private void enable_realtime() {
+        this.state = provider_state.ACTIVE;
+        GBApplication.deviceService().onEnableRealtimeSteps(true);
+        GBApplication.deviceService().onEnableRealtimeHeartRateMeasurement(true);
+
+        punchTimer.scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                LOG.debug("punching the deviceService...");
+                // As seen in LiveActivityFragment:
+                // have to enable it again and again to keep it measureing
+                GBApplication.deviceService().onEnableRealtimeHeartRateMeasurement(true);
+            }
+
+        }, 1000 * 10, 1000);
+        // Start after 10 seconds, repeat each second
+    }
+
+    @NonNull
+    private Cursor realtime(String[] projection, String[] args) {
```

Comment:
From looking at the name and the arguments, I would have no idea what it is or what arguments it expects. Maybe name it getRealtimeHeartRateCursor()?